ee/tabby-webserver/src/schema/auth.rs

```diff
@@ -354,6 +370,15 @@ pub trait AuthenticationService: Send + Sync {
         first: Option<usize>,
         last: Option<usize>,
     ) -> Result<Vec<InvitationNext>>;
+
+    async fn upsert_oauth_credential(
+        &self,
+        provider: OAuthProvider,
+        client_id: String,
+        client_secret: String,
+    ) -> Result<OAuthCredential>;
+    async fn delete_oauth_credential(&self, provider: OAuthProvider) -> Result<()>;
+    async fn get_oauth_credential(&self, provider: OAuthProvider) -> Result<OAuthCredential>;
```

Comment:
I’ll suggest skip the graphql api for now. So you can complete the end to end authorization flow by:
1. implementing the oauth callback.
2. set credentials by editing the sqlite database directly.

Once we finished 1 & 2, we can get back and review the graphql design.

Reply:
For the callback workflow, I've added new api
`/oauth/github/callback?code=ba327a......04a14db8`
and it produces following response:
```
{
    "access_token": "gho_G4ar.................1nmqYM",
    "scope": "user:email",
    "token_type": "bearer",
    "error": "",
    "error_description": "",
    "error_uri": ""
}
```

Is this supposed to expose to frontend? or we should fetch user info with the access token on the server side? @wsxiaoys

Reply:
As suggested above, I feel the github callback route can just use DbConn.

Reply:
It shouldn't be passed to frontend, instead, we should generate a redirect request to `/auth/signin?refreshToken=XXXXX` to finish signin process (the frontend will only see the new refreshtoken)

More comments here: https://github.com/TabbyML/tabby/pull/1160/files#r1442595004